datafusion/core/src/physical_plan/sorts/sort.rs

```diff
@@ -55,6 +55,9 @@ use tempfile::NamedTempFile;
 use tokio::sync::mpsc::{Receiver, Sender};
 use tokio::task;
 
+/// How much memory to reserve for performing in-memory sorts
+const EXTERNAL_SORTER_MERGE_RESERVATION: usize = 10 * 1024 * 1024;
```

Comment:
I'm not a massive fan of this, but this somewhat patches around the issue that once we initiate a merge we can't then spill

Reply:
The problem with this approach is that even 10MB may not be enough to correctly merge the batches prior to spilling. So some queries that today would succeed (though exceed their memory limits) might fail.

It seems to me better approaches (as follow on PRs) would be:
1. Make this a config parameter so users can avoid the error by reserving more memory up front if needed
2. teach SortExec how to write more (smaller) spill files if it doesn't have enough memory to merge the in memory batches. 

However, given the behavior on master today is to simply ignore the reservation and exceed the memory limit this behavior seems better than before. 
 
I suggest we merge this PR as is and file a follow on ticket for the improved behavior